src/registries/menu_items_registry.ts

```diff
@@ -24,89 +25,117 @@ import { SpreadsheetChildEnv } from "../types/env";
+  const name = item.name;
+  const children = item.children;
+  return {
+    id: item.id || new UuidGenerator().uuidv4(),
+    name: typeof name === "function" ? name : () => name,
+    isVisible: item.isVisible ? item.isVisible : () => true,
+    isEnabled: item.isEnabled ? item.isEnabled : () => true,
+    action: item.action,
+    children: children
+      ? (env) => {
+          return children
+            .map((child) => (typeof child === "function" ? child(env) : child))
+            .flat()
+            .map(createMenuItem);
+        }
+      : () => [],
+    isReadonlyAllowed: item.isReadonlyAllowed || false,
+    separator: item.separator || false,
+    icon: item.icon,
+    description: item.description || "",
+    textColor: item.textColor,
+    sequence: item.sequence || 0,
+  };
 }
 
 /**
  * The class Registry is extended in order to add the function addChild
  *
  */
-export class MenuItemRegistry extends Registry<FullMenuItem> {
+export class MenuItemRegistry extends Registry<MenuItemSpec> {
   /**
    * @override
    */
-  add(key: string, value: MenuItem): MenuItemRegistry {
-    this.content[key] = createFullMenuItem(key, value);
+  add(key: string, value: MenuItemSpec): MenuItemRegistry {
+    if (value.id === undefined) {
+      value.id = key;
+    }
+    this.content[key] = value;
     return this;
   }
   /**
    * Add a subitem to an existing item
    * @param path Path of items to add this subitem
    * @param value Subitem to add
    */
-  addChild(key: string, path: string[], value: MenuItem | MenuItemsBuilder): MenuItemRegistry {
+  addChild(key: string, path: string[], value: MenuItemSpec | MenuItemsBuilder): MenuItemRegistry {
+    if (typeof value !== "function" && value.id === undefined) {
+      value.id = key;
+    }
     const root = path.splice(0, 1)[0];
-    let node: FullMenuItem | undefined = this.content[root];
+    let node: MenuItemSpec | undefined = this.content[root];
     if (!node) {
       throw new Error(`Path ${root + ":" + path.join(":")} not found`);
     }
     for (let p of path) {
-      node = node.children.filter(isMenuItem).find((elt) => elt.id === p);
+      const children = node.children;
+      if (!children || typeof children === "function") {
+        throw new Error("bla bla");
+      }
+      node = children.find((elt) => elt.id === p);
 
       if (!node) {
         throw new Error(`Path ${root + ":" + path.join(":")} not found`);
       }
     }
-    if (typeof value !== "function") {
-      node.children.push(createFullMenuItem(key, value));
-    } else {
-      node.children.push(value);
+    if (!node.children) {
+      node.children = [];
     }
+    node.children.push(value);
     return this;
   }
 
-  /**
-   * Get a list of all elements in the registry, ordered by sequence
-   * @override
-   */
-  getAll(): FullMenuItem[] {
-    return super.getAll().sort((a, b) => a.sequence - b.sequence);
+  getMenu(): MenuItem[] {
```

Comment:
We could go one step further, integrate the isVisible here ?
```
getMenu(env: SpreadsheetChildEnv) {
  return createMenu(this.getAll()).filter((x) => x.isVisible(env));
}
```
We could also add this step in `createMenuItem` in order to avoid to create a menu item which is invisible

Reply:
Also, I'm not fan of the name `getMenu`, `getMenuItems` is not a better choice ?

Reply:
> We could go one step further, integrate the isVisible here ?
> 
> ```
> getMenu(env: SpreadsheetChildEnv) {
>   return createMenu(this.getAll()).filter((x) => x.isVisible(env));
> }
> ```
> 
> We could also add this step in `createMenuItem` in order to avoid to create a menu item which is invisible

I thought about something similar. But actually, I don't think it's a good idea to call `isVisible` too early (or any other function property). It should be called when we really need it and when we render the menu.
Because it might be visible at one time, but not 30s later (or the other way around).
Also, that way, we have a clear separation between the static menu definition and then the "runtime" menu where the properties have to be computed 

> Also, I'm not fan of the name getMenu, getMenuItems is not a better choice ?

isn't a "group of menu items" a "menu" ? :stuck_out_tongue: 
I don't mind changing the name though

Reply:
> I thought about something similar. But actually, I don't think it's a good idea to call isVisible too early (or any other function property). It should be called when we really need it and when we render the menu.
Because it might be visible at one time, but not 30s later (or the other way around).
Also, that way, we have a clear separation between the static menu definition and then the "runtime" menu where properties have to be computed

Ok, it's better to have a clear separation. It's just a pity that we have to apply the filter manually each time.

> isn't a "group of menu items" a "menu" ? 😛
I don't mind changing the name though

`getMenu` does not suggest me that it's an array that we get, so at first glance I had to go and see the definition. But I agree that a menu is a group of menu items, so ok for me

Reply:
> And for the filter how about handling it in the Menu component ? We handle the isEnabled there, I don't see why we coulnd't handle the isVisible 

👍